### PR TITLE
factory: cleaning up factory APIs to allow code sharing

### PR DESCRIPTION
**Pull request by alyssawilk**
Clean up inspired by #17745 

Risk Level: low (interface refactor)
Testing: n/a
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

File: source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc [hunk 1]
```diff
@@ -16,23 +16,24 @@ namespace Common {
 namespace DynamicForwardProxy {
 
 DnsCacheImpl::DnsCacheImpl(
-    Event::Dispatcher& main_thread_dispatcher, ThreadLocal::SlotAllocator& tls,
-    Random::RandomGenerator& random, Filesystem::Instance& file_system, Runtime::Loader& loader,
-    Stats::Scope& root_scope, ProtobufMessage::ValidationVisitor& validation_visitor,
+    Server::Configuration::FactoryContextBase& context,
     const envoy::extensions::common::dynamic_forward_proxy::v3::DnsCacheConfig& config)
-    : main_thread_dispatcher_(main_thread_dispatcher),
+    : main_thread_dispatcher_(context.dispatcher()),
```
Comment (jmarantz) [inline source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc:21]: Suggest renaming FactoryContextBase::dispatcher() to FactoryContextBase::mainThreadDispatcher to make it clearer that there can be lots of dispatchers but this is the one for the main thread.

**Review by jmarantz — COMMENTED** [review]
Nice refactor. One minor comment and I'm good to go.

File: source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc (hunk 1 above)
Comment (alyssawilk) [reply source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc:21]: Happy to do this but would you be Ok with a follow-up?  I suspect it's going to be significant churn.

Comment (jmarantz) [reply source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc:21]: sgtm wasn't aware there would be that much, but this is fine.

**Review by jmarantz — APPROVED** [review]

**Pull request merged**